File: native/android/core/src/main/java/com/vocare/translate/core/model/Languages.kt
package com.vocare.translate.core.model

/**
 * The language table, as on VOCA `main`: Cantonese (`yue`) was withdrawn from
 * the picker and flags were added. `yue` is still resolvable through
 * [Language.of] so history recorded before the withdrawal renders correctly.
 */
data class Language(val code: String, val flag: String, val label: String) {
    companion object {
        val ALL: List<Language> = listOf(
            Language("en", "🇺🇸", "English"),
            Language("zh", "🇨🇳", "Mandarin"),
            Language("ja", "🇯🇵", "Japanese"),
            Language("ko", "🇰🇷", "Korean"),
            Language("es", "🇪🇸", "Spanish"),
            Language("fr", "🇫🇷", "French"),
            Language("de", "🇩🇪", "German"),
            Language("ar", "🇦🇪", "Arabic"),
            Language("hi", "🇮🇳", "Hindi"),
            Language("fil", "🇵🇭", "Filipino (Tagalog)"),
        )

        private val WITHDRAWN = Language("yue", "🇭🇰", "Cantonese")

        /** Never null: an unknown code renders as itself with a globe, like the JSX `getLang`. */
        fun of(code: String?): Language {
            val c = code.orEmpty()
            return ALL.firstOrNull { it.code == c }
                ?: WITHDRAWN.takeIf { it.code == c }
                ?: Language(c, "🌐", c)
        }
    }
}

/**
 * Locale strings for on-device speech, per CONTRACT §4. Recognition and
 * synthesis are separate because Android's downloadable Mandarin recognition
 * pack answers only to `cmn-Hans-CN`, while ML Kit / TTS want `zh-CN`.
 */
object SpeechLocales {
    private val SPEECH = mapOf(
        "en" to "en-US", "zh" to "zh-CN", "yue" to "zh-HK", "ja" to "ja-JP", "ko" to "ko-KR",
        "es" to "es-ES", "fr" to "fr-FR", "de" to "de-DE", "ar" to "ar-SA", "hi" to "hi-IN", "fil" to "fil-PH",
    )
    private val ANDROID_RECOGNITION = mapOf("zh" to "cmn-Hans-CN")

    fun speech(code: String): String = SPEECH[code] ?: code
    fun recognition(code: String): String = ANDROID_RECOGNITION[code] ?: speech(code)
}

File: native/ios/VocaKit/Sources/VocaKit/App/Languages.swift
import Foundation

/// The JSX `LANGUAGES` table with flags. VOCA `main` withdrew Cantonese from
/// the picker (it has no on-device model anywhere and the cloud path is
/// Mandarin-only in practice), so `pickerLanguages` omits `yue` while `all`
/// keeps it so history rows written before the change still label correctly.
public struct Language: Identifiable, Equatable, Hashable, Sendable {
    public let code: String
    public let flag: String
    public let label: String

    public var id: String { code }

    public init(code: String, flag: String, label: String) {
        self.code = code
        self.flag = flag
        self.label = label
    }
}

public enum Languages {
    public static let all: [Language] = [
        Language(code: "en", flag: "🇺🇸", label: "English"),
        Language(code: "zh", flag: "🇨🇳", label: "Mandarin"),
        Language(code: "yue", flag: "🇭🇰", label: "Cantonese"),
        Language(code: "ja", flag: "🇯🇵", label: "Japanese"),
        Language(code: "ko", flag: "🇰🇷", label: "Korean"),
        Language(code: "es", flag: "🇪🇸", label: "Spanish"),
        Language(code: "fr", flag: "🇫🇷", label: "French"),
        Language(code: "de", flag: "🇩🇪", label: "German"),
        Language(code: "ar", flag: "🇦🇪", label: "Arabic"),
        Language(code: "hi", flag: "🇮🇳", label: "Hindi"),
        Language(code: "fil", flag: "🇵🇭", label: "Filipino (Tagalog)"),
    ]

    /// What the picker offers (VOCA main: no Cantonese).
    public static let pickerLanguages: [Language] = all.filter { $0.code != "yue" }

    /// JSX `getLang`: unknown codes fall back to a globe.
    public static func get(_ code: String) -> Language {
        all.first { $0.code == code } ?? Language(code: code, flag: "🌐", label: code)
    }

    /// Languages with a first-party on-device translation model on iOS (CONTRACT §4).
    public static let offlineCodes: [String] = ["en", "zh", "ja", "ko", "es", "fr", "de", "ar", "hi"]

    /// Whether this pair could be translated with no network on iOS. Says nothing
    /// about whether the packs are installed — `OfflineAvailability` answers that.
    public static func canTranslateOffline(_ a: String, _ b: String) -> Bool {
        !a.isEmpty && !b.isEmpty && a != b && offlineCodes.contains(a) && offlineCodes.contains(b)
    }

    /// JSX `LangScreen` search: label or code contains the query, case-insensitive.
    public static func search(_ query: String) -> [Language] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !q.isEmpty else { return pickerLanguages }
        return pickerLanguages.filter { $0.label.lowercased().contains(q) || $0.code.lowercased().contains(q) }
    }
}

// MARK: - Formatting helpers (JSX formatMinutes / formatDuration / formatClock)

public enum VocaFormat {
    /// Tier chip: `12 min` at ≥10 minutes, else `m:ss`.
    public static func minutes(_ seconds: Int) -> String {
        let clamped = max(0, seconds)
        let mins = clamped / 60
        let secs = clamped % 60
        if mins >= 10 { return "\(mins) min" }
        return "\(mins):" + String(format: "%02d", secs)
    }

    /// History rows: `3m 12s` or `45s`.
    public static func duration(_ seconds: Int?) -> String {
        guard let seconds else { return "" }
        let m = seconds / 60
        let s = seconds % 60
        return m > 0 ? "\(m)m \(s)s" : "\(s)s"
    }

    /// Live-bar clock: tabular MM:SS, never negative.
    public static func clock(_ seconds: Int) -> String {
        let n = max(0, seconds)
        return String(format: "%02d:%02d", n / 60, n % 60)
    }
}
